examples/next-app-router/src/app/page.tsx

```diff
@@ -0,0 +1,22 @@
+import { HydrateClient, trpc } from '~/trpc/server';
+import { Suspense } from 'react';
+import { Posts } from './_components/posts';
+import { UserButton } from './_components/user-button';
+
+export default async function Home() {
+  // prefetch in RSC
+  await trpc.post.list();
+
+  return (
+    <main className="container max-w-xl py-8">
+      <Suspense>
+        <div className="absolute right-8 top-8">
+          <UserButton />
+        </div>
+      </Suspense>
+      <HydrateClient>
```

Comment:
good that this is contextual

if you have many of these, i guess they keep on sending stuff doubly? it would stack up quite a lot of it was used a lot

maybe we could keep track of stuff that's already been sent to the client so we don't send same query times across many hydration boundaries

Reply:
> if you have many of these, i guess they keep on sending stuff doubly? it would stack up quite a lot of it was used a lot

i would actually have thought RQ would handle this internally. isn't this an issue no matter where you use them?

Reply:
there is no deduplication with `HydrationBoundary`. Every boundary will send its content to the client.

@ephem we talked about this some time ago I think ?

Reply:
I think there is if you use the same `queryClient` maybe ?

Reply:
it should be a single query client per request https://github.com/trpc/trpc/pull/5451/files#diff-8ef8de9fb13401a32e846aeaedf32d659dd5f222706751573fb792d04d36e66cR24

Reply:
I think we've moved away from this recommendation precisely because of this:

https://tanstack.com/query/v5/docs/framework/react/guides/advanced-ssr#alternative-use-a-single-queryclient-for-prefetching

> In the example above, we create a new queryClient for each Server Component that fetches data. This is the recommended approach, but if you want to, you can alternatively create a single one that is reused across all Server Components:

also:

> As a future improvement, we might look into creating a dehydrateNew() function (name pending) that only dehydrate queries that are new since the last call to dehydrateNew(). Feel free to get in touch if this sounds interesting and like something you want to help out with!

Is this you getting in touch to implement that feature 😅 ?

Reply:
aha ;)